src/Interpreters/RewriteCountDistinctVisitor.cpp

```diff
@@ -0,0 +1,75 @@
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Interpreters/RewriteCountDistinctVisitor.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Common/typeid_cast.h>
+#include "Coordination/KeeperStorage.h"
+#include "Parsers/ASTExpressionList.h"
+#include "Parsers/ASTIdentifier.h"
+#include "Parsers/ASTSelectQuery.h"
+#include "Parsers/ASTSubquery.h"
+#include "Parsers/ASTTablesInSelectQuery.h"
+#include <Parsers/Lexer.h>
+#include <Parsers/parseQuery.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTSelectWithUnionQuery.h>
+
+namespace DB
+{
+
+void RewriteCountDistinctFunctionMatcher::visit(ASTPtr & ast, Data & /*data*/)
+{
+    auto * selectq = ast->as<ASTSelectQuery>();
+    if (!selectq || !selectq->tables() || selectq->tables()->children.size() != 1)
+        return;
+    auto expr_list = selectq->select();
+    if (!expr_list || expr_list->children.size() != 1)
+        return;
+    auto * func = expr_list->children[0]->as<ASTFunction>();
+    if (!func || (Poco::toLower(func->name) != "countdistinct" && Poco::toLower(func->name) != "uniqexact"))
+        return;
+    auto arg = func->arguments->children;
+    if (arg.size() != 1)
+        return;
+    if (!arg[0]->as<ASTIdentifier>())
+        return;
+    auto * table_expr = selectq->tables()->as<ASTTablesInSelectQuery>()->children[0]->as<ASTTablesInSelectQueryElement>()->children[0]->as<ASTTableExpression>();
```

Comment:
let's also check that `selectq->tables()->as<ASTTablesInSelectQuery>()->children[0]->as<ASTTablesInSelectQueryElement>()->children.size() == 1`